File: backend/app/schemas/historical.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import InvoiceStatus
from app.schemas.common import ORMModel


class HistoricalMappingIn(BaseModel):
    boq_line_number: int | None = None
    boq_item_id: UUID | None = None
    description: str | None = None
    quantity: Decimal = Field(ge=0)
    amount: Decimal = Field(ge=0)


class HistoricalInvoiceIn(BaseModel):
    invoice_number: str = Field(min_length=1, max_length=120)
    vendor_trn: str | None = None
    vendor_id: UUID | None = None
    contract_number: str | None = None
    contract_id: UUID | None = None
    invoice_date: date
    subtotal: Decimal = Field(ge=0)
    vat: Decimal = Field(ge=0)
    total: Decimal = Field(gt=0)
    status: str = "unpaid"
    paid_amount: Decimal = Field(default=Decimal("0"), ge=0)
    payment_date: date | None = None
    payment_reference: str | None = None
    mappings: list[HistoricalMappingIn] = []


class HistoricalPreviewRow(BaseModel):
    invoice_number: str
    vendor_trn: str
    contract_number: str
    invoice_date: date
    subtotal: Decimal
    vat: Decimal
    total: Decimal
    status: str
    paid_amount: Decimal
    payment_date: date | None
    payment_reference: str | None
    errors: list[str] = []


class HistoricalMappingPreview(BaseModel):
    invoice_number: str
    boq_line_number: int
    quantity: Decimal
    amount: Decimal
    errors: list[str] = []


class HistoricalPreviewOut(BaseModel):
    invoices: list[HistoricalPreviewRow]
    mappings: list[HistoricalMappingPreview]
    row_errors: int
    unresolved_contracts: list[str] = []
    unresolved_vendors: list[str] = []
    unresolved_boq_lines: list[str] = []


class HistoricalCommitRequest(BaseModel):
    invoices: list[HistoricalInvoiceIn]


class HistoricalInvoiceOut(ORMModel):
    id: UUID
    contract_id: UUID | None
    vendor_id: UUID | None
    invoice_number: str
    invoice_date: date
    subtotal: Decimal
    vat: Decimal
    total: Decimal
    currency: str
    status: InvoiceStatus
    archived: bool
    superseded_by_id: UUID | None
    created_at: datetime


class HistoricalEditRequest(BaseModel):
    invoice_date: date | None = None
    subtotal: Decimal | None = None
    vat: Decimal | None = None
    total: Decimal | None = None
    status: str | None = None
    paid_amount: Decimal | None = None
    payment_date: date | None = None
    payment_reference: str | None = None
    mappings: list[HistoricalMappingIn] | None = None

File: frontend/src/api/historical.ts
import { api } from "@/api/client";

export type HistoricalInvoice = {
  id: string;
  contract_id: string | null;
  vendor_id: string | null;
  invoice_number: string;
  invoice_date: string;
  subtotal: string;
  vat: string;
  total: string;
  currency: string;
  status: string;
  archived: boolean;
  superseded_by_id: string | null;
  created_at: string;
};

export type HistoricalPreviewRow = {
  invoice_number: string;
  vendor_trn: string;
  contract_number: string;
  invoice_date: string;
  subtotal: string;
  vat: string;
  total: string;
  status: string;
  paid_amount: string;
  payment_date: string | null;
  payment_reference: string | null;
  errors: string[];
};

export type HistoricalMappingPreview = {
  invoice_number: string;
  boq_line_number: number;
  quantity: string;
  amount: string;
  errors: string[];
};

export type HistoricalPreview = {
  invoices: HistoricalPreviewRow[];
  mappings: HistoricalMappingPreview[];
  row_errors: number;
  unresolved_contracts: string[];
  unresolved_vendors: string[];
  unresolved_boq_lines: string[];
};

export type HistoricalMappingIn = {
  boq_line_number?: number | null;
  quantity: number | string;
  amount: number | string;
};

export type HistoricalInvoiceIn = {
  invoice_number: string;
  vendor_trn?: string | null;
  contract_number?: string | null;
  invoice_date: string;
  subtotal: number | string;
  vat: number | string;
  total: number | string;
  status?: string;
  paid_amount?: number | string;
  payment_date?: string | null;
  payment_reference?: string | null;
  mappings?: HistoricalMappingIn[];
};

export async function previewHistorical(
  file: File,
  mappingsFile?: File | null,
): Promise<HistoricalPreview> {
  const form = new FormData();
  form.append("file", file);
  if (mappingsFile) form.append("mappings_file", mappingsFile);
  return api<HistoricalPreview>("/historical-invoices/preview", { form });
}

export const commitHistorical = (
  invoices: HistoricalInvoiceIn[],
): Promise<HistoricalInvoice[]> =>
  api<HistoricalInvoice[]>("/historical-invoices/commit", { body: { invoices } });

export const listHistorical = (contractId?: string): Promise<HistoricalInvoice[]> =>
  api<HistoricalInvoice[]>(
    contractId ? `/historical-invoices?contract_id=${contractId}` : "/historical-invoices",
  );

export const deleteHistorical = (id: string): Promise<void> =>
  api<void>(`/historical-invoices/${id}`, { method: "DELETE" });
